src/compiler/scala/tools/nsc/transform/patmat/Logic.scala

```diff
@@ -752,17 +752,15 @@ trait ScalaLogic extends Interface with Logic with TreeAndTypeAnalysis {
 
       private val uniques = new mutable.HashMap[Type, Const]
       private[TreesAndTypesDomain] def unique(tp: Type, mkFresh: => Const): Const =
-        uniques.get(tp).getOrElse(
-          uniques.find {case (oldTp, oldC) => oldTp =:= tp} match {
-            case Some((_, c)) =>
-              debug.patmat("unique const: "+ ((tp, c)))
-              c
-            case _ =>
-              val fresh = mkFresh
-              debug.patmat("uniqued const: "+ ((tp, fresh)))
-              uniques(tp) = fresh
-              fresh
+        uniques.getOrElse(tp, {
+          // normalize to increase the chance of structural equality and reduce the cost of =:=
+          val normalized = tp.map(_.normalize)
```

Comment:
Why not tp.normalize?

Reply:
`normalize` is shallow - it will only dealias the top-level type alias right?

Reply:
Yes indeed:
```scala
scala> type A = Int
type A

scala> type F[x] = List[x]
type F

scala> typeOf[F[A]].normalize
val res0: $r.intp.global.Type = List[A]
```